From f6e3e3918e54a9490165a7ded770840583a0124f Mon Sep 17 00:00:00 2001
From: Thang Nguyen <thang.nguyen.ue@renesas.com>
Date: Fri, 20 Sep 2024 09:40:59 +0700
Subject: [PATCH] Support small resolutions

This patch consists of 2 parts:

1. Edit work-around supporting allocate output buffer in Bypass:
  Reuse gst_omx_video_dec_configure_bypass for support small
  resolutions, rename it to gst_omx_video_dec_create_out_pool. Because
  from OMX v.1.3.0, Bypass mode of OMX can raise initial
  OMX_EventPortSettingsChanged. It's mean Gstreamer can reconfigure
  output buffer dynamically and the work-around can be removed.

2. Support small resolutions (reuse part of above work-around):
  * Change OMX raise OMX_EventPortSettingsChanged threshold to minimum
    value instead of default value.
  * For inputs with resolution smaller than minimum value, OMX will not
    raise OMX_EventPortSettingsChanged, resulting in Gstreamer not
    allocating OMX output buffer. So it needs to be done manually. Use
    gst_omx_video_dec_create_out_pool because it has that function.

Signed-off-by: Thang Nguyen <thang.nguyen.ue@renesas.com>
---
 omx/gstomxvideodec.c | 46 ++++++++++++++++++++++++++++++++++----------
 1 file changed, 36 insertions(+), 10 deletions(-)

diff --git a/omx/gstomxvideodec.c b/omx/gstomxvideodec.c
index e09129d..3eae3aa 100644
--- a/omx/gstomxvideodec.c
+++ b/omx/gstomxvideodec.c
@@ -108,6 +108,10 @@ enum
 #define GST_OMX_VIDEO_DEC_INTERNAL_ENTROPY_BUFFERS_DEFAULT (5)
 #define GST_OMX_VIDEO_DEC_NUMBER_OUTPUT_BUFFERS_DEFAULT    (0)
 #define GST_OMX_VIDEO_DEC_NUMBER_OUTPUT_BUFFERS_MAXIMUM    (32)
+#define GST_OMX_VIDEO_DEC_MIN_FRAMEWIDTH                   (80)
+#define GST_OMX_VIDEO_DEC_MIN_FRAMEHEIGHT                  (80)
+#define GST_OMX_VIDEO_DEC_MIN_STRIDE                       (96)
+#define GST_OMX_VIDEO_DEC_MIN_SLICEHEIGHT                  (80)
 
 /* class initialization */
 
@@ -1690,9 +1694,9 @@ done:
   return err;
 }
 
-/* Allocate and configure output buffers for Bypass mode */
+/* Create a new out port pool */
 static OMX_ERRORTYPE
-gst_omx_video_dec_configure_bypass (GstOMXVideoDec * self)
+gst_omx_video_dec_create_out_pool (GstOMXVideoDec * self)
 {
   GstOMXPort *port;
   OMX_ERRORTYPE err;
@@ -2749,6 +2753,8 @@ static gboolean
 gst_omx_video_dec_enable (GstOMXVideoDec * self, GstBuffer * input)
 {
   GstOMXVideoDecClass *klass = GST_OMX_VIDEO_DEC_GET_CLASS (self);
+  OMX_U32 min_stride = GST_OMX_VIDEO_DEC_MIN_STRIDE;
+  OMX_U32 min_sliceheight = GST_OMX_VIDEO_DEC_MIN_SLICEHEIGHT;
 
   GST_DEBUG_OBJECT (self, "Enabling component");
 
@@ -2811,10 +2817,19 @@ gst_omx_video_dec_enable (GstOMXVideoDec * self, GstBuffer * input)
       if (!gst_omx_video_dec_allocate_in_buffers (self))
         return FALSE;
 
+      /* Round up to get minimum stride and slice height follow bypass 
+       * alignment */
       if (self->bypass) {
-        /* In G2L Bypass mode, allocate output buffers here instead
-         * of waiting for Event PortSettingChanged */
-        if (gst_omx_video_dec_configure_bypass(self))
+        min_stride = GST_ROUND_UP_128(GST_OMX_VIDEO_DEC_MIN_FRAMEWIDTH);
+        min_sliceheight = GST_ROUND_UP_32(GST_OMX_VIDEO_DEC_MIN_FRAMEHEIGHT);
+      }
+      /* If the input resolution is smaller than the minimum stride and slice
+       * height, no output pool will be created. Therefore, the output pool
+       * needs to be created manually. */
+      if ((self->dec_in_port->port_def.format.video.nFrameWidth <= min_stride)
+          && (self->dec_in_port->port_def.format.video.nFrameHeight
+              <= min_sliceheight)) {
+        if (gst_omx_video_dec_create_out_pool (self) != OMX_ErrorNone)
           return FALSE;
       } else {
         if (gst_omx_port_allocate_buffers (self->dec_out_port) != OMX_ErrorNone)
@@ -2915,6 +2930,8 @@ gst_omx_video_dec_set_format (GstVideoDecoder * decoder,
   gboolean needs_disable = FALSE;
   OMX_PARAM_PORTDEFINITIONTYPE port_def;
   OMX_U32 framerate_q16 = gst_omx_video_calculate_framerate_q16 (info);
+  OMX_U32 min_stride = GST_OMX_VIDEO_DEC_MIN_STRIDE;
+  OMX_U32 min_sliceheight = GST_OMX_VIDEO_DEC_MIN_SLICEHEIGHT;
 
   self = GST_OMX_VIDEO_DEC (decoder);
   klass = GST_OMX_VIDEO_DEC_GET_CLASS (decoder);
@@ -3088,21 +3105,30 @@ gst_omx_video_dec_set_format (GstVideoDecoder * decoder,
     OMX_PARAM_PORTDEFINITIONTYPE out_port_def;
 
     /* Initialize default output allocation align for page size
-     * Choose 192x192 for VP9 Decoder and 128x128 for other Decoders
+     * Choose 192x192 for VP9 Decoder and 96x80 for other Decoders
      * because they are close to minimum setting 192 (VP9) and 80 (others)*/
     gst_omx_port_get_port_definition (self->dec_out_port, &out_port_def);
     if (GST_IS_OMX_VP9_DEC (self)) {
       out_port_def.format.video.nStride = 192;
       out_port_def.format.video.nSliceHeight = 192;
     } else {
+      /* Round up to get minimum stride and slice height follow bypass 
+       * alignment */
       if (self->bypass) {
-        /* In G2L Bypass mode, OMX will not send Event PortSettingChanged
-         * so application has to set parameters for output port manually */
+        min_stride = GST_ROUND_UP_128(GST_OMX_VIDEO_DEC_MIN_FRAMEWIDTH);
+        min_sliceheight = GST_ROUND_UP_32(GST_OMX_VIDEO_DEC_MIN_FRAMEHEIGHT);
+      }
+      /* If the input resolution is smaller than the minimum stride and slice
+       * height, the output will not be reconfigured. Therefore, the output size
+       * needs to be manually adjusted instead of taking the default value. */
+      if ((info->width <= min_stride) && (info->height <= min_sliceheight)) {
         out_port_def.format.video.nFrameWidth =  info->width;
         out_port_def.format.video.nFrameHeight =  info->height;
-        out_port_def.format.video.nStride = (info->width+127)/128*128;
-        out_port_def.format.video.nSliceHeight = (info->height+15)/16*16;
       }
+      /* To OMX can raise OMX_EventPortSettingsChanged if the input resolution
+       * is larger than the minimum stride and slice height. */
+      out_port_def.format.video.nStride = min_stride;
+      out_port_def.format.video.nSliceHeight = min_sliceheight;
     }
 
     if (gst_omx_port_update_port_definition (self->dec_out_port,
-- 
2.25.1

